src/main/kotlin/org/jitsi/nlj/util/PacketCache.kt

```diff
@@ -16,85 +16,129 @@
-    private class RtpPacketCache(
-        timeout: Duration,
-        maxNumElements: Int
-    ) {
-        private val cache = TimeExpiringCache<Int, RtpPacket>(timeout, maxNumElements)
-        private val rfc3711IndexTracker = Rfc3711IndexTracker()
-
-        fun insert(packet: RtpPacket) {
-            val index = rfc3711IndexTracker.update(packet.sequenceNumber)
-            cache.insert(index, packet)
+    fun updateTimestamp(ssrc: Long, seqNum: Int, timeAdded: Long) = getCache(ssrc).updateTimestamp(seqNum, timeAdded)
+}
+
+/**
+ * Implements a cache for RTP packets.
+ */
+class RtpPacketCache(
+    size: Int,
+    synchronize: Boolean = true,
+    timeProvider: TimeProvider = TimeProvider()
+) : ArrayCache<RtpPacket>(size, synchronize, { it.clone() }, timeProvider) {
+
+    private val rfc3711IndexTracker = Rfc3711IndexTracker()
+
+    override fun discardItem(item: RtpPacket) {
+        BufferPool.returnBuffer(item.buffer)
+    }
+
+    /**
+     * Gets a packet with a given RTP sequence number from the cache.
+     */
+    fun get(sequenceNumber: Int): Container? {
+        // Note that we use [interpret] because we don't want the ROC to get out of sync because of funny requests
+        // (NACKs)
+        val index = rfc3711IndexTracker.interpret(sequenceNumber)
+        return super.getContainer(index)
+    }
+
+    fun insert(rtpPacket: RtpPacket): Boolean {
+        val index = rfc3711IndexTracker.update(rtpPacket.sequenceNumber)
+        return super.insertItem(rtpPacket, index)
+    }
+
+    fun getMany(numBytes: Int): Set<RtpPacket> =
+        if (synchronize) {
+            synchronized (syncRoot) {
+                doGetMany(numBytes)
+            }
+        } else {
+            doGetMany(numBytes)
         }
 
-        fun get(seqNum: Int): RtpPacket? {
-            // We don't know to which ROC value this sequence number belongs, so we'll try
-            // to find it first using the current ROC value, but if that fails we'll try
-            // again to use the previous one, just in case we've rolled over recently
-            return cache.get(seqNum + rfc3711IndexTracker.roc * 0x1_0000) ?:
-                cache.get(seqNum + rfc3711IndexTracker.roc - 1 * 0x1_0000)
+    fun updateTimestamp(seqNum: Int, timeAdded: Long) {
+        val index = rfc3711IndexTracker.interpret(seqNum)
+        super.updateTimeAdded(index, timeAdded)
+    }
+
+    fun doGetMany(numBytes: Int): Set<RtpPacket> {
+        var bytesRemaining = numBytes
+        val packets = mutableSetOf<RtpPacket>()
+
+        val lastIndex = lastIndex
+        if (lastIndex == -1) {
+            return packets
         }
 
-        fun getMany(numBytes: Int): Set<RtpPacket> {
-            var bytesRemaining = numBytes
-            val packets = mutableSetOf<RtpPacket>()
-            cache.forEachDescending { pkt ->
-                packets.add(pkt)
-                bytesRemaining -= pkt.length
-                bytesRemaining > 0
+        for (i in 0 until size) {
```

Comment:
wondering if maybe it's worth adding a method which does the equivalent of `forEachDescending`: basically add a helper to hide away the logic of iterating through the cached packets from newest to oldest?

Reply:
Yep, tt's much better this way